### PR TITLE
Rebase dbaas-refactor for merging to upstream main

### PR DESCRIPTION
**Pull request by jianrongzhang89**
This PR is to rebase dbaas-reafactor code for DBaaS work, squash the commits and fix go lint issues found, as the preparation for submitting a PR for merging to main branch of the mongodb upstream. Documentation is added to README.md as well.

The new operator has been deployed on the lab and tested successfully.

### CODE REVIEW

File: bundle.Dockerfile [hunk 1]
```diff
@@ -1,20 +1,21 @@
 FROM scratch
-LABEL com.redhat.openshift.versions="v4.5-v4.7"
-LABEL com.redhat.delivery.backport=true
-LABEL com.redhat.delivery.operator.bundle=true
```
Comment (tchughesiv) [inline bundle.Dockerfile:4]: these seem like they could be important?

File: config/manager/kustomization.yaml [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
-#  - name: manager-config
-#    files:
-#      - controller_manager_config.yaml
\ No newline at end of file
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: controller
+  newName: quay.io/ecosystem-appeng/mongodb-atlas-operator
+  newTag: 0.6.25
```
Comment (tchughesiv) [inline config/manager/kustomization.yaml:11]: should this section be included in PR to mongo?

File: test.txt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+=== RUN   TestDBaaSProviderCreate
+=== RUN   TestDBaaSProviderCreate/Nominal
+=== RUN   TestDBaaSProviderCreate/CRDCheckFail
+=== RUN   TestDBaaSProviderCreate/CRDCheckNotFound
+--- PASS: TestDBaaSProviderCreate (0.01s)
+    --- PASS: TestDBaaSProviderCreate/Nominal (0.01s)
+    --- PASS: TestDBaaSProviderCreate/CRDCheckFail (0.00s)
+    --- PASS: TestDBaaSProviderCreate/CRDCheckNotFound (0.00s)
+PASS
+ok  	github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/dbaasprovider	(cached)
```
Comment (tchughesiv) [inline test.txt:54]: is this file intentional?

File: bundle.Dockerfile (hunk 1 above)
Comment (jeremyary) [reply bundle.Dockerfile:4]: Also our DBaaS UX only works on 4.8+ so we should make sure that's included

Comment (jianrongzhang89) [reply bundle.Dockerfile:4]: OK. I added them back and include 4.8

File: config/manager/kustomization.yaml (hunk 2 above)
Comment (jianrongzhang89) [reply config/manager/kustomization.yaml]: OK. it is updated.

File: test.txt (hunk 3 above)
Comment (jianrongzhang89) [reply test.txt]: Not really. I cleaned it up.

File: config/crd/bases/atlas.mongodb.com_atlasclusters.yaml [hunk 4]
```diff
@@ -166,7 +166,6 @@ spec:
                     type: string
                 required:
                 - name
-                - namespace
```
Comment (jeremyary) [inline config/crd/bases/atlas.mongodb.com_atlasclusters.yaml:169]: have we changed something that's inadvertently affected these Atlas models (this and atlasdatabaseusers) this way?

File: config/crd/bases/atlas.mongodb.com_atlasdatabaseusers.yaml [hunk 5]
```diff
@@ -91,7 +91,6 @@ spec:
                     type: string
                 required:
                 - name
-                - namespace
```
Comment (jeremyary) [inline config/crd/bases/atlas.mongodb.com_atlasdatabaseusers.yaml:94]: same here, curious how we've modified their existing models

File: config/crd/bases/atlas.mongodb.com_atlasclusters.yaml (hunk 4 above)
Comment (jianrongzhang89) [reply config/crd/bases/atlas.mongodb.com_atlasclusters.yaml:169]: Well, this is the result of code generation, since namespace is set here in the type definition:
https://github.com/mongodb/mongodb-atlas-kubernetes/blob/main/pkg/api/v1/common.go#L23

What do you think we should do?

Comment (jianrongzhang89) [reply config/crd/bases/atlas.mongodb.com_atlasclusters.yaml:169]: As discussed we will keep it as it is.

File: config/crd/bases/atlas.mongodb.com_atlasdatabaseusers.yaml (hunk 5 above)
Comment (jianrongzhang89) [reply config/crd/bases/atlas.mongodb.com_atlasdatabaseusers.yaml:94]: Same as above

**Review by redhatHameed — APPROVED** [review]

**Review by jeremyary — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -117,4 +117,214 @@ containers:
...
+    icon:
+      base64data: <icon image data>
+      mediatype: image/png
+    name: Red Hat DBaaS / MongoDB Atlas
+```
+If the Atlas Operator is undeployed with the OLM, the above registration CR gets cleaned up automatically.
+
+**2.** Check MongoDBAtlasInventory Customer Resource
+
+When an application developer creates a [DBaaSInventory](https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/config/crd/bases/dbaas.redhat.com_dbaasinventories.yaml) CR for MongoDB, the DBaaS Operator creates a MongoDBAtlasInventory CR, and the Atlas Operator discovers the clusters and  instances, and sets the result in the CR status.
```
Comment (jeremyary) [inline README.md:167]: this would be an administrator instead of an application developer

File: README.md [hunk 7]
```diff
@@ -117,4 +117,214 @@ containers:
...
+  - lastTransitionTime: "2021-08-18T20:06:47Z"
+    message: Spec sync OK
+    reason: SyncOK
+    status: "True"
+    type: SpecSynced
+  instances:
+  - instanceID: 608df625aa94426b4167e45f
+    instanceInfo:
+      connectionStringsStandard: ""
+      connectionStringsStandardSrv: mongodb+srv://dbaas-cluster1.h38mo.mongodb.net
```
Comment (jeremyary) [inline README.md:200]: please redact any live URLs or credentials

File: README.md [hunk 8]
```diff
@@ -117,4 +117,214 @@ containers:
...
+      instanceSizeName: M10
+      projectID: 60b798fea37f9f09acc1a154
+      projectName: mytest
+      providerName: AWS
+      regionName: US_EAST_1
+    name: test
+```
+**3.** Check MongoDBAtlasConnection Custom Resource
+
+The application developer can then create a [DBaaSConnection](https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/config/crd/bases/dbaas.redhat.com_dbaasconnections.yaml) CR for connection to the MongoDB database instace found, the DBaaS Operator automatically creates a MongoDBAtlasConnection CR. The Atlas Operator creates a database user in Atlas for the cluster with the default database `admin`. The Atlas Operator stores the db user corredentials in a kubernetes secret, and the remaining connection information in a configmap, and then updates the MongoDBAtlasConnection CR status.
```
Comment (jeremyary) [inline README.md:240]: s/instace/instance
s/corredentials/credentials

File: README.md [hunk 9]
```diff
@@ -117,4 +117,214 @@ containers:
...
+    name: atlas-connection-cm-knp9z
+  credentialsRef:
+    name: atlas-db-user-5pc8b
+```
+The corresponding generated secret:
+```
+apiVersion: v1
+data:
+  password: cGFzczEyM3dvcmRA
+  username: ZGJVc2VyXzEwMQ==
```
Comment (jeremyary) [inline README.md:281]: please redact any live URLs or credentials - these are an easy decode away from giving the world a target URL & credentials to access it!

**Review by jeremyary — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -117,4 +117,214 @@ containers:
...
+      connectionStringsStandardSrv: mongodb+srv://dbaas-cluster1.h38mo.mongodb.net
+      instanceSizeName: M0
+      projectID: 608df5e652e1944293e8584a
+      projectName: Project 1
+      providerName: TENANT
+      regionName: US_EAST_1
+    name: DBaaS-Cluster1
+  - instanceID: 60807282b4ab8d3b3c84be53
+    instanceInfo:
+      connectionStringsStandard: mongodb://test-shard-00-00.iojsa.mongodb.net:27017,test-shard-00-01.iojsa.mongodb.net:27017,test-shard-00-02.iojsa.mongodb.net:27017/?ssl=true&authSource=admin&replicaSet=atlas-133pql-shard-0
```
Comment (redhatHameed) [inline README.md:209]: are not change this to show as a host  and isSrv field ?, also not sure we need to show the connection details here

File: README.md (hunk 9 above)
Comment (jianrongzhang89) [reply README.md:281]: Done!

File: README.md (hunk 8 above)
Comment (jianrongzhang89) [reply README.md:240]: Corrected

File: README.md (hunk 7 above)
Comment (jianrongzhang89) [reply README.md:200]: Done

File: README.md (hunk 6 above)
Comment (jianrongzhang89) [reply README.md:167]: OK

File: README.md (hunk 10 above)
Comment (jianrongzhang89) [reply README.md]: Removed connectionStringsStandard as it is not used currently. connectionStringsStandardSrv is kept in the inventory as it makes a little issue for troubleshooting.

**Review by redhatHameed — APPROVED** [review]

File: README.md [hunk 11]
```diff
@@ -117,4 +117,198 @@ containers:
...
+```
+### Step 4. Test Database as a Service (DBaaS) on OpenShift
+
+The Atlas Operator is integrated with the [Red Hat Database-as-a-Service (DBaaS) Operator](https://github.com/RHEcosystemAppEng/dbaas-operator) which allows application developers to import database instances and connect to the databases through the [Service Binding Operator](https://github.com/redhat-developer/service-binding-operator). More information can be found [here](https://github.com/RHEcosystemAppEng/dbaas-operator#readme).
+
+Note that both the DBaaS Operator and Atlas Operator should be installed through the [Operator Lifecyle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager).
+
+**1.** Check DBaaS Registration
+
+If the DBaaS Operator has been deployed in the OpenShift Cluster, the Atlas Operator automatically creates a cluster level [DBaaSProvider](https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/config/crd/bases/dbaas.redhat.com_dbaasproviders.yaml) customer resource (CR) object `mongodb-atlas-registration` to automatically register itself with the DBaaS Operator.
```
Comment (jeremyary) [inline README.md:129]: s/customer/custom missed this one earlier, sorry :)

File: README.md [hunk 12]
```diff
@@ -117,4 +117,198 @@ containers:
...
+    name: atlas-connection-cm-knp9z
+  credentialsRef:
+    name: atlas-db-user-5pc8b
+```
+The corresponding generated secret:
+```
+apiVersion: v1
+data:
+  password: cGFzczEyM3dAcmQ=
+  username: ZGJVc2VyXzEwMQ==
```
Comment (jeremyary) [inline README.md:271]: I still feel like these should be redacted here, even though we're redacting the URL where they could be used... if that password value, which is a very easy decode,is used as a default on user creation or something this is not good to push out.

File: README.md (hunk 11 above)
Comment (jianrongzhang89) [reply README.md]: Corrected

File: README.md (hunk 12 above)
Comment (jianrongzhang89) [reply README.md:271]: These are already redacted with dummy values: dbUser_101/pass123w@rd.

Comment (jeremyary) [reply README.md:271]: ok thanks, I decoded the values and the pw seemed like a legit one, so just wanted to make sure.

File: README.md [hunk 13]
```diff
@@ -2,7 +2,7 @@
 [![MongoDB Atlas Operator](https://github.com/mongodb/mongodb-atlas-kubernetes/workflows/Test/badge.svg)](https://github.com/mongodb/mongodb-atlas-kubernetes/actions/workflows/test.yml?query=branch%3Amain)
 [![MongoDB Atlas Go Client](https://img.shields.io/badge/Powered%20by%20-go--client--mongodb--atlas-%2313AA52)](https://github.com/mongodb/go-client-mongodb-atlas)
 
-The MongoDB Atlas Operator provides a native integration between the Kubernetes orchestration platform and MongoDB Atlas — the only multi-cloud document database service that gives you the versatility you need to build sophisticated and resilient applications that can adapt to changing customer demands and market trends.
+The MongoDB Atlas Operator provides a native integration between the Kubernetes orchestration platform and MongoDB Atlas — the only multi-cloud document database service that gives you the versatility you need to build sophisticated and resilient applications that can adapt to changing dbaas-cluster1.a00aa.omeromer demands and market trends.
```
Comment (jeremyary) [inline README.md:5]: I don't think `dbaas-cluster1.a00aa.omeromer` belongs in here. Accidental paste maybe?

Comment (jianrongzhang89) [reply README.md]: My bad. Corrected

**Review by jeremyary — APPROVED** [review]
